### ISSUE TITLE
bug: userland modules preferred over core modules

### ISSUE THREAD

**Issue by boneskull** [rollup/rollup#544]
[Reference](https://nodejs.org/api/modules.html#modules_all_together)

I *think* this is a rollup-plugin-node-resolve issue, but I could be wrong, which is why it's here.

Try something like this:

1. Create a file `entry.js`:

  ```js
  import events from 'events';

  const ee = new events.EventEmitter();

  ee.emit('foo');

  export default ee;
  ```

2. Execute `npm install events rollup-plugin-commonjs rollup-plugin-node-resolve`.  Ensure `rollup` is installed globally or locally as well.

3. Create `rollup.config.js`:

  ```js
  import cjs from 'rollup-plugin-commonjs';
  import npm from 'rollup-plugin-node-resolve';
  export default {
    entry: 'entry.js',
    plugins: [
      npm({jsnext: true, main: true}),
      cjs({include: 'node_modules/**'})
    ]
  };
  ```

4.  Execute `rollup -c rollup.config.js > out.js`

5.  Inspect the resulting file.  It's the main entry point of the userland `events` module; not the core `events` module (note lack of `usingDomains` property or `init` function).

What especially sucks about this is that when using npm 3, your deps are flattened.  So, if you depend on a package *X* which depends on userland `events`, you will get userland `events` whether you like it or not--even if package *X* is a `devDependency`.

**Comment by boneskull** [comment]
I *think* a workaround may be to blacklist things.  I understand `import` !== `require`.  Regardless, I don't believe it should prefer a userland module if it purports to support cjs/node.

**Comment by boneskull** [comment]
or just use `require('events')`, unfortunately

**Comment by TrySound** [comment]
@boneskull All builtin module names skipped by default.

**Comment by ljharb** [comment]
@boneskull this actually seems correct - isn't rollup like browserify, which means you'd only use it on a bundle you deliver to a browser? core modules can't be delivered to a browser.

**Comment by boneskull** [comment]
@trysound @ljharb if rollup's intent is to limit itself to the browser, then that would be true.  If that's its intent, then projects like [compiled](https://www.npmjs.com/package/compiled) are using it erroneously.  

Can this behavior be disabled via a flag?

**Comment by boneskull** [comment]
cc @bahmutov

**Comment by eventualbuddha** [comment]
This is due to a bug in node-resolve, as linked above. We can work around this in rollup-plugin-node-resolve.

**Comment by Rich-Harris** [comment]
This is fixed in rollup-plugin-node-resolve@1.5.0 – thanks

**Comment by boneskull** [comment]
@TrySound Was there a misunderstanding or were you just explaining the current behavior?

@Rich-Harris Thanks!
